# Coding Exercises

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Time Series Data in Pandas: Introduction


### Exercise 1: Intro to Time Series Data & DatetimeIndex

In [1]:
#run the cell!
import pandas as pd

1. __Import__ the csv-file "__aapl_fb.csv__" containing Stock Price Data for Apple and Facebook for the years 2015-2018! Convert the datatype in the column __Date__ to __datetime__ and set the column __Date__ as the __Index__ of the DataFrame! <br>
__Save__ the DataFrame in the variable __stocks__! __Fill in the gaps!__ 

In [82]:
#There was no time component, so parse_dates=["Datetime"] failed, looked at soln for this one.
stocks = pd.read_csv("aapl_fb.csv", parse_dates=["Date"], index_col=["Date"])
result = (len(stocks.index), stocks.FB.max())
result

(1005, 217.5)

How many Timestamps does __stocks__ contain? What was the __maximum stock price__ for Facebook in that Period?

In [83]:
# run the cell!
stocks.head()

AAPL         FB
Date                             
2015-01-02  109.330002  78.449997
2015-01-05  106.250000  77.190002
2015-01-06  106.260002  76.150002
2015-01-07  107.750000  76.150002
2015-01-08  111.889999  78.180000

In [84]:
# run the cell!
stocks.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1005 entries, 2015-01-02 to 2018-12-28
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   AAPL    1005 non-null   float64
 1   FB      1005 non-null   float64
dtypes: float64(2)
memory usage: 23.6 KB


In [85]:
# run the cell!
stocks.describe()

AAPL           FB
count  1005.000000  1005.000000
mean    140.988338   133.415502
std      35.082112    35.846003
min      90.339996    74.050003
25%     112.120003   105.019997
50%     129.360001   129.070007
75%     168.630005   167.410004
max     232.070007   217.500000

There are in total ... 1005 Timestamps with stock price data. Facebooks´s maximum stock price was ... 217.50!

2. __Select__ the stock prices on __May 13, 2016__! What was the Price for __Apple__?

In [87]:
stocks.loc["2016-MAY-13"]

AAPL     90.519997
FB      119.809998
Name: 2016-05-13 00:00:00, dtype: float64

The Apple Stock Price on that day was ... 90.52

3. __Select__ all stock prices in __September 2017__ and calculate the __mean__ for both stocks! What was the __mean stock price__ for __Apple__ in __September 2017__?

Apple´s monthly mean price for September 2017 is ... 157.61! 

4. __Create a Timestamp__ for August 10, 2019! (convert __2 alternative date strings__ of your choice) 

In [ ]:
# Alt 1


In [ ]:
# Alt 2


5. __Change Frequency__ of stocks to __monthly__ ("Downsampling") and display the __last__ traded stock price per month! (Hint: The Index should show the last calender day of each month) <br>
What is the __last__ traded price for __Facebook__ in __March 2015__?

The last traded price is ... 82.22!

6. Calculate __monthly average/mean stock prices__ and save the new DataFrame in the variable __stocks_m__!

What was the average stock price for __Facebook__ in __April 2015__?

In [ ]:
# run the cell!
stocks_m.head()

The average price in april is ... 82.09!

7. __Downsample__ stocks to __weekly__ frequency, showing prices on __Fridays__!
What was __Facebook´s__ stock price on the __fourth Friday__ in __January 2015__?

Facebook´s stock price on the fourth Friday in January 2015 was ... 77.83!

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## Time Series Data in Pandas: Introduction


### Exercise 1: Intro to Time Series Data & DatetimeIndex

In [ ]:
#run the cell!
import pandas as pd

1. __Import__ the csv-file "__aapl_fb.csv__" containing Stock Price Data for Apple and Facebook for the years 2015-2018! Convert the datatype in the column __Date__ to __datetime__ and set the column __Date__ as the __Index__ of the DataFrame! <br>
__Save__ the DataFrame in the variable __stocks__! __Fill in the gaps!__ 

In [ ]:
stocks = pd.read_csv("aapl_fb.csv", parse_dates= ["Date"], index_col = "Date")

How many Timestamps does __stocks__ contain? What was the __maximum stock price__ for Facebook in that Period?

In [ ]:
# run the cell!
stocks.head()

In [ ]:
# run the cell!
stocks.info()

In [ ]:
# run the cell!
stocks.describe()

There are in total ... 1005 Timestamps with stock price data. Facebooks´s maximum stock price was ... 217.50!

2. __Select__ the stock prices on __May 13, 2016__! What was the Price for __Apple__?

In [ ]:
stocks.loc["2016-05-13"]

The Apple Stock Price on that day was ... 90.52

3. __Select__ all stock prices in __September 2017__ and calculate the __mean__ for both stocks! What was the __mean stock price__ for __Apple__ in __September 2017__?

In [ ]:
stocks.loc["2017-09"].mean()

Apple´s monthly mean price for September 2017 is ... 157.61! 

4. __Create a Timestamp__ for August 10, 2019! (convert __2 alternative date strings__ of your choice) 

In [ ]:
# Alt 1
pd.to_datetime("2019-08-10")

In [ ]:
# Alt 2
pd.to_datetime("10 Aug 2019")

5. __Change Frequency__ of stocks to __monthly__ ("Downsampling") and display the __last__ traded stock price per month! (Hint: The Index should show the last calender day of each month) <br>
What is the __last__ traded price for __Facebook__ in __March 2015__?

In [ ]:
stocks.resample("M").last()

The last traded price is ... 82.22!

6. Calculate __monthly average/mean stock prices__ and save the new DataFrame in the variable __stocks_m__!

In [ ]:
stocks_m  = stocks.resample("M").mean()

What was the average stock price for __Facebook__ in __April 2015__?

In [ ]:
# run the cell!
stocks_m.head()

The average price in april is ... 82.09!

7. __Downsample__ stocks to __weekly__ frequency, showing prices on __Fridays__!
What was __Facebook´s__ stock price on the __fourth Friday__ in __January 2015__?

In [ ]:
stocks.resample("W-Fri").last()

Facebook´s stock price on the fourth Friday in January 2015 was ... 77.83!

# Well Done!

-------------------